src/utilities/resolveSchemaCoordinate.ts

```diff
@@ -0,0 +1,190 @@
+      readonly enumValue: GraphQLEnumValue;
+    }
+  | {
+      readonly kind: 'FieldArgument';
+      readonly type: GraphQLNamedType;
+      readonly field: GraphQLField<unknown, unknown>;
+      readonly fieldArgument: GraphQLArgument;
+    }
+  | {
+      readonly kind: 'Directive';
+      readonly directive: GraphQLDirective;
+    }
+  | {
+      readonly kind: 'DirectiveArgument';
+      readonly directive: GraphQLDirective;
+      readonly directiveArgument: GraphQLArgument;
+    };
+
+/**
+ * A schema coordinate is resolved in the context of a GraphQL schema to
+ * uniquely identifies a schema element. It returns undefined if the schema
+ * coordinate does not resolve to a schema element.
+ *
+ * https://spec.graphql.org/draft/#sec-Schema-Coordinates.Semantics
+ */
+export function resolveSchemaCoordinate(
+  schema: GraphQLSchema,
+  schemaCoordinate: string | Source,
+): ResolvedSchemaElement | undefined {
+  return resolveASTSchemaCoordinate(
+    schema,
+    parseSchemaCoordinate(schemaCoordinate),
+  );
+}
+
+/**
+ * Resolves schema coordinate from a parsed SchemaCoordinate node.
+ */
+export function resolveASTSchemaCoordinate(
+  schema: GraphQLSchema,
+  schemaCoordinate: SchemaCoordinateNode,
+): ResolvedSchemaElement | undefined {
+  const { ofDirective, name, memberName, argumentName } = schemaCoordinate;
+  if (ofDirective) {
+    // SchemaCoordinate :
+    //   - @ Name
+    //   - @ Name ( Name : )
+    // Let {directiveName} be the value of the first {Name}.
+    // Let {directive} be the directive in the {schema} named {directiveName}.
+    const directive = schema.getDirective(name.value);
+    if (!argumentName) {
+      // SchemaCoordinate : @ Name
+      // Return the directive in the {schema} named {directiveName}.
+      if (!directive) {
+        return;
+      }
+      return { kind: 'Directive', directive };
+    }
+
+    // SchemaCoordinate : @ Name ( Name : )
+    // Assert {directive} must exist.
+    if (!directive) {
+      return;
+    }
+    // Let {directiveArgumentName} be the value of the second {Name}.
+    // Return the argument of {directive} named {directiveArgumentName}.
+    const directiveArgument = directive.args.find(
+      (arg) => arg.name === argumentName.value,
+    );
+    if (!directiveArgument) {
+      return;
+    }
+    return { kind: 'DirectiveArgument', directive, directiveArgument };
+  }
+
+  // SchemaCoordinate :
+  //   - Name
+  //   - Name . Name
+  //   - Name . Name ( Name : )
+  // Let {typeName} be the value of the first {Name}.
+  // Let {type} be the type in the {schema} named {typeName}.
+  const type = schema.getType(name.value);
+  if (!memberName) {
+    // SchemaCoordinate : Name
+    // Return the type in the {schema} named {typeName}.
+    if (!type) {
+      return;
+    }
+    return { kind: 'NamedType', type };
+  }
+
+  if (!argumentName) {
+    // SchemaCoordinate : Name . Name
+    // If {type} is an Enum type:
+    if (isEnumType(type)) {
```

Comment:
Depending what we decide about meta fields (see https://github.com/graphql/graphql-spec/pull/794/files#r1900996045), we might want to assert that `memberName.value` does not start with `__` here.

Reply:
resolving this conversatin, we have decided to allow meta-fields